### PR TITLE
Skaffold run support new parameter '--render-output' with '--render-only

### PR DESCRIPTION
**Pull request by kissmikijr**
Relates to https://github.com/GoogleContainerTools/skaffold/issues/3582

Should merge before : n/a

Should merge after : n/a

**Description**
Added new parameter for skaffold dev and run. --output, currently it works only with --render-only. If --render-only and --output specified the output of --render-only will be written into the filepath given to --output
**User facing changes**
skaffold run got a new parameter
**Before**

n/a

**After**

n/a

**Next PRs.**

n/a

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Mentions any output changes.
- [x] Adds documentation as needed: user docs, YAML reference, CLI reference.
- [ ] Adds integration tests if needed.

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit test added.
- [ ] User facing changes look good.

**Release Notes**

```
Adds a new flag `--render-output` to write rendered output to a file.
```

### CODE REVIEW

**Comment by kissmikijr** [comment]
This is a pretty quick fix for the issue. I am not sure if it is the best, I wanted to do something like where --render-only could get a filepath and if it is not specified it prints out the results.

Regarding to unittesting this feature or throwing a nice friendly message if someone tries to use --output without --render-only I would appreciate some guidance where to find them.

Thanks!

File: cmd/skaffold/app/cmd/flags.go [hunk 1]
```diff
@@ -227,6 +227,15 @@ var FlagRegistry = []Flag{
...
 		DefinedOn:     []string{"dev", "run"},
 	},
+	{
+		Name:          "output",
+		Shorthand:     "o",
+		Usage:         "Writes '--render-only' output to the specified file",
+		Value:         &opts.Output,
+		DefValue:      "",
+		FlagAddMethod: "StringVar",
+		DefinedOn:     []string{"dev", "run"},
```
Comment (balopat) [inline cmd/skaffold/app/cmd/flags.go]: does this flag really need to be defined on `dev` as well?

File: cmd/skaffold/app/cmd/flags.go [hunk 2]
```diff
@@ -227,6 +227,15 @@ var FlagRegistry = []Flag{
 		FlagAddMethod: "BoolVar",
 		DefinedOn:     []string{"dev", "run"},
 	},
+	{
+		Name:          "output",
```
Comment (balopat) [inline cmd/skaffold/app/cmd/flags.go]: I'd name this as `--render-output` to clarify the scope in the name of the flag too

File: docs/content/en/schemas/v2alpha3.json [hunk 3]
```diff
@@ -494,6 +494,15 @@
           "description": "builder image used.",
           "x-intellij-html-description": "builder image used."
         },
+        "buildpacks": {
```
Comment (balopat) [inline docs/content/en/schemas/v2alpha3.json]: I'm surprised this came up here. Can you rerun `hack/generate-schemas.sh`?

File: pkg/skaffold/build/buildpacks/lifecycle.go [hunk 4]
```diff
@@ -62,12 +62,13 @@ func (b *Builder) build(ctx context.Context, out io.Writer, a *latest.Artifact,
...
-		AppPath:  workspace,
-		Builder:  artifact.Builder,
-		RunImage: artifact.RunImage,
-		Env:      envMap(env),
-		Image:    latest,
-		NoPull:   alreadyPulled,
+		AppPath:    workspace,
+		Builder:    artifact.Builder,
+		RunImage:   artifact.RunImage,
+		Buildpacks: artifact.Buildpacks,
```
Comment (balopat) [inline pkg/skaffold/build/buildpacks/lifecycle.go]: again, irrelevant to your change - not sure how this showed up here.

File: pkg/skaffold/config/options.go [hunk 5]
```diff
@@ -50,6 +50,7 @@ type SkaffoldOptions struct {
 	AutoSync           bool
 	AutoDeploy         bool
 	RenderOnly         bool
+	Output             string
```
Comment (balopat) [inline pkg/skaffold/config/options.go]: I'd rename this here too: `RenderOutput`

File: pkg/skaffold/schema/latest/config.go [hunk 6]
```diff
@@ -697,6 +697,11 @@ type BuildpackArtifact struct {
 	// For example: `["key1=value1", "key2=value2", "key3={{.ENV_VARIABLE}}"]`.
 	Env []string `yaml:"env,omitempty"`
 
+	// Buildpacks is a list of strings, where each string is a specific buildpack to use with the builder.
+	// If you specify buildpacks the builder image automatic detection will be ignored. These buildpacks will be used to build the Image from your source code.
+	// Order matters.
+	Buildpacks []string `yaml:"buildpacks,omitempty"`
```
Comment (balopat) [inline pkg/skaffold/schema/latest/config.go]: we should remove all the buildpacks related stuff from this PR.

File: cmd/skaffold/app/cmd/flags.go (hunk 1 above)
Comment (kissmikijr) [reply cmd/skaffold/app/cmd/flags.go]: fixed.

File: cmd/skaffold/app/cmd/flags.go (hunk 2 above)
Comment (kissmikijr) [reply cmd/skaffold/app/cmd/flags.go]: Fixed

File: docs/content/en/schemas/v2alpha3.json (hunk 3 above)
Comment (kissmikijr) [reply docs/content/en/schemas/v2alpha3.json]: Fixed, I did something wrong with the branch creating, sorry

File: pkg/skaffold/config/options.go (hunk 5 above)
Comment (kissmikijr) [reply pkg/skaffold/config/options.go]: Fixed

File: pkg/skaffold/schema/latest/config.go (hunk 6 above)
Comment (kissmikijr) [reply pkg/skaffold/schema/latest/config.go]: Done.

**Comment by balopat** [comment]
Kokoro failure is due to known flake: https://github.com/GoogleContainerTools/skaffold/issues/3302

**Comment by balopat** [comment]
One last thing - we should add a test for this

**Comment by kissmikijr** [comment]
Yes! I'm on it, thank you for the help :)

Balint Pato <notifications@github.com> ezt írta (időpont: 2020. jan. 30.,
Cs, 1:48):

**Comment by balopat** [comment]
There's something weird going on with the commits appearing here and the sudden increase in changed file numbers. I didn't have time to look at it, but can you try to rebase your own commits on top of master? Let me know if you're stuck. The CLA status is failing as all the imported commits now have 6-7 authors altogether who would need to give their consent. That will be fixed as soon as we fix the commits.

**Comment by kissmikijr** [comment]
@balopat I am trying to fix the issues, but I am stuck, any help would be appreciated.

**Comment by kissmikijr** [comment]
@balopat I think i fixed the commits, could you please help me with the test? Where should I put it, I wasn't able to figure it out. I was trying to make something in the deploy_test.go, but i am not sure if that is the correct place. 
Other question, how should I test this behaviour? I was thinking about something like to expect the r.Render to be called with the correct parameters if both '--render-only' and '--render-output' got specified.
thanks for the help

**Comment by nkubala** [comment]
@kissmikijr sorry for the delayed response here. there are two places you could add tests:
* `integration/render_test.go` - add a test that does a render with the new flag, read that newly written file and verify it against a hardcoded skaffold.yaml string. you can follow the pattern of the test that's in there.
* `pkg/skaffold/deploy/{kubectl, kustomize, helm}_test.go` - very similar to the integration test, but with the commands overwritten. you can check the exact flags passed to the command to make sure you're getting the new flag piped all the way through.

let me know if you need any more help with that!

**Comment by AndiDog** [comment]
For writing tests, maybe look at, or wait for, the PR #3567 – I've invested quite some time to get a nice integration test working where the `skaffold render` command is executed directly.

**Comment by kissmikijr** [comment]
@AndiDog awesme! thanks, nice work, I'm gonna for it to be merged!

**Comment by geritol** [comment]
@googlebot I consent.

**Comment by geritol** [comment]
Hi @nkubala!

I have added the integration test for render, see 897aea2

I did not really understand the second part. 
I have added an integration test for run 5f5b98b, but not sure on how should assert. Should I apply the manifests to the namespace and assert the same way as the tests that are run on gcp? Or should I partially check the generated file's content?

**Comment by briandealwis** [comment]
@kissmikijr it looks like there is a test failure: https://travis-ci.com/github/GoogleContainerTools/skaffold/jobs/298296133#L1154

**Comment by kissmikijr** [comment]
@briandealwis Hey, yeah, we have a question about what/how should we test this. If you could check out @geritol 's comment above and could guide us a little bit We'd appreciate that! :) 
Thank you in advance.

**Comment by AndiDog** [comment]
https://github.com/GoogleContainerTools/skaffold/pull/3567/files#diff-d3f21d5008ea0ca75c6dfac3afbce9a4 has the test function `TestRenderFromBuildOutput` which runs the built skaffold binary to perform integration tests. Should be a good starting point. Mind that the PR I linked is still pending approval, so you have to check out the proposed branch to see the example integration test.

**Comment by nkubala** [comment]
@kissmikijr @geritol sorry for not getting back to you guys earlier! and thanks @AndiDog for helping provide some guidance :) 

the integration test for render looks good. for the integration test for run, I think checking the file contents to make sure it matches what's expected is probably good enough. let's not bother actually applying the rendered manifests to a namespace for now.

for the unit tests, what I meant to say was to add a test that checks to make sure that the correct calls are being made in skaffold when the `--render-only` flag is passed. so you can add a test case to https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/deploy/kubectl_test.go#L60, something like "render only", and make sure that it never calls a `kubectl apply` in the `commands` field of the test. hopefully that makes sense!

also, unfortunately this PR will need to be rebased quickly. let me know if there's anything else I can do to help get this merged!

File: integration/render_test.go [hunk 7]
```diff
@@ -28,6 +29,74 @@ import (
...
+
+	test := struct {
+		description string
+		builds      []build.Artifact
+		labels      []deploy.Labeller
+		renderPath  string
+		input       string
+		expectedOut string
+	}{
+		description: "it persist rendered manifest to provided filepath",
```
Comment (nkubala) [inline integration/render_test.go:45]:
```suggestion
		description: "write rendered manifest to provided filepath",
```

File: integration/run_test.go [hunk 8]
```diff
@@ -122,6 +124,39 @@ func TestRun(t *testing.T) {
 	}
 }
 
+func TestRunRenderOnly(t *testing.T) {
+	if testing.Short() || RunOnGCP() {
+		t.Skip("skipping kind integration test")
+	}
+
+	testutil.Run(t, "it persist rendered manifest to provided filepath", func(tu *testutil.T) {
```
Comment (nkubala) [inline integration/run_test.go:132]:
```suggestion
	testutil.Run(t, "writet rendered manifest to provided filepath", func(tu *testutil.T) {
```

**Comment by tejal29** [comment]
@kissmikijr ping!

**Comment by geritol** [comment]
@nkubala sorry for the late reply. Did the rebase, checking your suggestions regarding the tests

**Comment by geritol** [comment]
Started adding the test that @nkubala suggested, that makes sure deploy render-only does not call `kubectl apply`, but it seems to. Currently I am not sure why.
cc. @kissmikijr

**Comment by nkubala** [comment]
@geritol your test is failing because the test itself calls `k.Deploy` with the `KubectlDeployer` it instantiates. 

https://github.com/GoogleContainerTools/skaffold/blob/60b9a1e9ec2b6de4112420e7c4f922457bd35af6/pkg/skaffold/deploy/kubectl_test.go#L209

normally skaffold would check the `RenderOnly` option and conditionally call into `Render()` instead:

https://github.com/GoogleContainerTools/skaffold/blob/60b9a1e9ec2b6de4112420e7c4f922457bd35af6/pkg/skaffold/runner/deploy.go#L36-L38

but the test logic itself is bypassing that check. you'll need to either rewrite the test here to create a `SkaffoldRunner` and call `Deploy` on that object (instead of the embedded `Deployer` itself in the `Runner`), or write a different test that does that. does that make sense?

**Comment by geritol** [comment]
I think it does, I have tryed to add a test to `skaffold/pkg/skaffold/deploy/kubectl_test.go` that uses `SkaffoldRunner` but it fails because `import cycle not allowed in test`

**Comment by nkubala** [comment]
@geritol that error happens because you're introducing a circular dependency into the code - *usually* (but not always) a sign that you're trying to break the very loose dependency model that we have in the code. can you either restructure this test to operate on different structs, or move the logic to a different package?

if you can't do either of these, then maybe just write a new test in a different package. you're very close!

**Comment by geritol** [comment]
Ok, thanks. I got that, but I was not sure where this test could fit well.
I will try to find a place for it

**Comment by geritol** [comment]
Hey @nkubala I think I managed to put together a test that makes sure that no kubeclt calls are made when running deploy render-only, see https://github.com/GoogleContainerTools/skaffold/pull/3586/commits/34127b7dc0037c106536eaf1d9c9322126a82af6.

Not sure this is the best way to do it, so your input is welcome.

File: pkg/skaffold/runner/deploy_test.go [hunk 9]
```diff
@@ -133,3 +135,26 @@ func TestDeployNamespace(t *testing.T) {
...
+		}
+
+		r := SkaffoldRunner{
+			runCtx:     runCtx,
+			kubectlCLI: kubectl.NewFromRunContext(runCtx),
+			deployer:   getDeployer(runCtx),
+		}
+		var builds []build.Artifact
+
+		err := r.Deploy(context.Background(), ioutil.Discard, builds)
```
Comment (tejal29) [inline pkg/skaffold/runner/deploy_test.go:156]: @kissmikijr Thanks you rebasing and addressing the comments. Can you please explain how does this test ensure your are not making any kubectl calls?
Does this test point to a non existing `kubecontext` and hence should fail.

Comment (tejal29) [reply pkg/skaffold/runner/deploy_test.go:156]: if that is the case, can you please make it explicit by either adding a comment or 

```
KubeContext: "does-not-exist",
```

Comment (geritol) [reply pkg/skaffold/runner/deploy_test.go:156]: added `KubeContext: "does-not-exist"`

**Comment by kissmikijr** [comment]
hey @tejal29 If you have time could you check out the latest changes? 

We would really appreciate it, thank you!

**Comment by tejal29** [comment]
@kissmikijr Sorry for the delay!

**Comment by nkubala** [comment]
@geritol @kissmikijr sorry i missed the last review here, but thanks for all your continued efforts on this one!

**Pull request merged**